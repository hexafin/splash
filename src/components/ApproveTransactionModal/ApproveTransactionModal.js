import React, { Component } from "react";
import { View, Text, TouchableOpacity, StyleSheet, Image, Alert } from "react-native"
import { colors } from "../../lib/colors"
import LoadingCircle from "../universal/LoadingCircle"
import Checkmark from "../universal/Checkmark"
import TouchID from "react-native-touch-id"
import Button from "../universal/Button"
import { cryptoUnits, decimalToUnits, unitsToDecimal } from "../../lib/cryptos"
import { BITCOIN_ERRORS } from '../../bitcoin-api'

class ApproveTransactionModal extends Component {

	constructor(props) {
		super(props)
		this.state = {
			success: false,
			wasLoading: false,
			feeSatoshi: null,
		}
	}

	componentDidMount() {

		const {
			address,
			amount,
			currency,
		    exchangeRate
		} = this.props

	    const rate = decimalToUnits(exchangeRate, 'USD')
	    const satoshiAmount = (currency == 'USD') ? Math.round((amount/rate)*cryptoUnits.BTC) : amount

		api.GetBitcoinFees({network: this.props.bitcoinNetwork, from: this.props.userBitcoinAddress, amtSatoshi: satoshiAmount}).then(feeSatoshi => {
			this.setState(prevState => {
				return {
					...prevState,
					feeSatoshi,
				}
			})			
		})
	}

	componentWillReceiveProps(nextProps) { 
		if(nextProps.loading == true) {
			this.setState(prevState => {
				return {
					...prevState,
					wasLoading: true
				}
			})			
		}

		if(nextProps.loading == false && this.state.wasLoading == true) {
			this.setState(prevState => {
				return {
					...prevState,
					wasLoading: false,
					success: true,
				}
			})			
		}
	}

	render() {

		const {
			address,
			toId,
			toSplashtag,
			amount,
			currency,
		    exchangeRate,
		    biometricEnabled,
		} = this.props

	    const rate = decimalToUnits(exchangeRate, 'USD')
	    const satoshiAmount = (currency == 'USD') ? Math.round((amount/rate)*cryptoUnits.BTC) : amount
	    const relativeAmount = (currency == 'USD') ? amount : Math.round((amount/cryptoUnits.BTC) * rate)

	    let fee = 0
	    if (this.state.feeSatoshi) {
			fee = unitsToDecimal(this.state.feeSatoshi, 'BTC')
	    }
		const totalSatoshiAmount = satoshiAmount+this.state.feeSatoshi
		const totalRelativeAmount = Math.round((totalSatoshiAmount/cryptoUnits.BTC) * rate)

		const runTransaction = () => {
			this.props.SendTransaction(address, satoshiAmount, this.state.feeSatoshi, relativeAmount, toId, toSplashtag)
				.then(() => this.props.dismiss(true))
				.catch(error => {
					if (error == BITCOIN_ERRORS.BALANCE) {
						Alert.alert("Insufficient balance", null, [
										{text: "Dismiss", onPress: () => {
											this.props.dismiss(false)
										}}
									])
					} else if (error == BITCOIN_ERRORS.UTXOS) {
						Alert.alert("Insufficient available balance. Please wait for your transactions to be confirmed before sending more.", null, [
										{text: "Dismiss", onPress: () => {
											this.props.dismiss(false)
										}}
									])
					} else if (error == BITCOIN_ERRORS.FEE) {
						Alert.alert("Bitcoin fee is greater than balance. Try lowering the fee in settings.", null, [
										{text: "Dismiss", onPress: () => {
											this.props.dismiss(false)
										}}
									])
					}
				})

			// 5 second timeout	
			setTimeout(() => {
				if (this.props.loading) {
					this.props.showTimeoutModal()								
				}
			}, 5000)
		}

		const confirm = () => {
			if (totalSatoshiAmount) {
				if (biometricEnabled) {
					TouchID.authenticate("Sign transaction biometrically").then(() => {
						runTransaction()
					}).catch(error => {
						this.props.dismiss(false)
					})
				}
				else {
					runTransaction()
				}
			} else {
				Alert.alert("Unable to calculate fee. Please try again")
			}
		}

		return (
				<View style={styles.content}>
	                <View style={styles.header}>
	                  <Text style={styles.title}>Confirm transaction</Text>
	                  <TouchableOpacity onPress={() => this.props.dismiss(false)}>
	                    <Image style={styles.closeButton} source={require('../../assets/icons/Xbutton.png')}/>
	                  </TouchableOpacity>
	                </View>
	                <View style={styles.information}>
	                  <Text style={styles.exchangeText}>1 BTC = ${unitsToDecimal(rate, 'USD')} USD</Text>
	                  <View style={styles.amountBox}>
	                    <Text style={styles.relativeText}>USD ${unitsToDecimal(totalRelativeAmount, 'USD')}</Text>
	                    <Text style={styles.amountText}>{unitsToDecimal(totalSatoshiAmount, 'BTC')} BTC</Text>
	                  </View>
	                </View>
	                <View style={styles.feeBox}>
	                    <Text style={styles.description}>Blockchain fee —— {fee} BTC</Text>
	                    <Text style={styles.description}>They receive —— {unitsToDecimal(satoshiAmount, 'BTC')} BTC</Text>
	                </View>
	                <Text style={[styles.description, {paddingBottom: 15}]}>To {address}</Text>
	                  <Button
	                  	onPress={confirm}
	                  	style={styles.button} 
	                  	loading={this.props.loading && !this.state.success}
	                  	checkmark={this.state.success && !this.props.loading}
	                  	checkmarkPersist={true}
						checkmarkCallback={() => this.props.dismiss(true)}
						disabled={(this.props.error != null) ? true : false}
						title={"Send Transaction"} primary={true}/>
				</View>
		)
	}
}

export default ApproveTransactionModal

const styles = StyleSheet.create({
  content: {
    flex: 1,
    paddingTop: 25,
    paddingBottom: 65,
    paddingHorizontal: 25,
    justifyContent: 'space-between',
    alignSelf: "stretch",
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingBottom: 10,
  },
  closeButton: {
	height: 20,
	width: 20,
	margin: 8,
  },
  title: {
    fontSize: 20,
    fontWeight: "600",
    color: colors.nearBlack,
  },
  information: {
    flex: 1,
    paddingHorizontal: 32,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  exchangeText: {
	fontSize: 14,
	fontWeight: '500',
	color: colors.gray,
	paddingBottom: 5,
  },
  amountBox: {
    height: 83,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'stretch',
    borderRadius: 8,
    backgroundColor: '#6364F1',
  },
  relativeText: {
  	fontSize: 28,
  	color: colors.white,
  	fontWeight: '600'
  },
  amountText: {
  	fontSize: 18,
  	opacity: 0.77,
  	color: colors.white,
  	fontWeight: '600'
  },
  feeBox: {
  	padding: 10
  },
  description: {
    textAlign: 'center',
    fontSize: 12,
    color: colors.gray,
    fontWeight: '500',
  },
  footer: {
  	flexDirection: 'row',
  	paddingTop: 10,
  	alignSelf: 'center',
  	alignItems: 'center'
  },
  lockIcon: {
  	height: 13,
  	width: 10
  },
  securedText: {
  	paddingLeft: 10,
  	backgroundColor: 'rgba(0,0,0,0)',
  	color: colors.lightGray,
  	fontSize: 15,
  	fontWeight: '600'
  }
})