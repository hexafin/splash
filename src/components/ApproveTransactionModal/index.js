import ApproveTransactionModal from "./ApproveTransactionModal"
import {connect} from "react-redux"
import {bindActionCreators} from "redux"
import {SendTransaction, DismissTransaction} from "../../redux/transactions/actions"


const showTimeoutModal = () => {
  return {
    type: 'SHOW_MODAL',
    modalType: 'INFO',
    modalProps: {
      title: 'Sending timeout',
      body: 'Please connect to the internet and try again.',
    },   
  }
}

const mapStateToProps = (state) => {
    const activeCryptoCurrency = state.crypto.activeCryptoCurrency
    return {
      activeCryptoCurrency,
      loading: state.transactions.isSendingTransaction,
      error: state.transactions.errorSendingTransaction,
      network: state.crypto.wallets[activeCryptoCurrency].network,
      userAddress: state.crypto.wallets[activeCryptoCurrency].address,
      exchangeRate: state.crypto.exchangeRates[activeCryptoCurrency].USD,
      biometricEnabled: state.user.biometric,
    }
}

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
      SendTransaction, showTimeoutModal
    }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(ApproveTransactionModal)
