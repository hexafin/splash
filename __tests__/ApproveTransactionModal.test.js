import "react-native";
import React from "react";
import ApproveTransactionModal from "../src/components/ApproveTransactionModal/ApproveTransactionModal";

it("ApproveTransactionModal renders correctly", () => {
	const snap = shallow(
		<ApproveTransactionModal
			toAddress={"testAddress"}
			toId={"testId"}
			toSplashtag={"@testSplash"}
			amount={100}
			currency={"BTC"}
			exchangeRate={0.01}
			biometricEnabled={true}
			activeCryptoCurrency={"BTC"}
		/>
	);
	expect(snap).toMatchSnapshot();
});
