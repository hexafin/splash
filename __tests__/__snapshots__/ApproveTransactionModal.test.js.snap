// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApproveTransactionModal renders correctly 1`] = `
<Component
  style={
    Object {
      "alignSelf": "stretch",
      "flex": 1,
      "justifyContent": "space-between",
      "paddingBottom": 65,
      "paddingHorizontal": 25,
      "paddingTop": 25,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 10,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#333333",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
    >
      Confirm transaction
    </Component>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Component
        source={
          Object {
            "testUri": "../../../src/assets/icons/Xbutton.png",
          }
        }
        style={
          Object {
            "height": 20,
            "margin": 8,
            "width": 20,
          }
        }
      />
    </TouchableOpacity>
  </Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-around",
        "paddingHorizontal": 32,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#919191",
          "fontSize": 14,
          "fontWeight": "500",
          "paddingBottom": 5,
        }
      }
    >
      1 
      BTC
       = $
      0.01
       USD
    </Component>
    <Component
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "#6364F1",
          "borderRadius": 8,
          "height": 83,
          "justifyContent": "center",
        }
      }
    >
      <Component
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 28,
            "fontWeight": "600",
          }
        }
      >
        USD $
        0.00
      </Component>
      <Component
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "600",
            "opacity": 0.77,
          }
        }
      >
        0.00000
         
        BTC
      </Component>
    </Component>
  </Component>
  <Component
    style={
      Object {
        "padding": 10,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#919191",
          "fontSize": 11,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      Blockchain fee —— 
      0
       
      BTC
       ($
      0
      )
    </Component>
    <Component
      style={
        Object {
          "color": "#919191",
          "fontSize": 11,
          "fontWeight": "500",
          "textAlign": "center",
        }
      }
    >
      They receive —— 
      0.00000
       
      BTC
       ($
      0.00
      )
    </Component>
  </Component>
  <Component
    style={
      Array [
        Object {
          "color": "#919191",
          "fontSize": 11,
          "fontWeight": "500",
          "textAlign": "center",
        },
        Object {
          "paddingBottom": 15,
        },
      ]
    }
  >
    To 
    testAddress
  </Component>
  <Button
    checkmark={false}
    checkmarkCallback={[Function]}
    checkmarkPersist={true}
    disabled={false}
    onPress={[Function]}
    primary={true}
    title="Send Transaction"
  />
</Component>
`;
